internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -624,6 +570,61 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 	return CompositionResult{ConnectionDetails: d.GetComposite().GetConnectionDetails(), Composed: resources, Events: events, Conditions: conditions}, nil
 }
 
+func (c *FunctionComposer) runFunction(ctx context.Context, fn v1.PipelineStep, req *v1beta1.RunFunctionRequest) (*v1beta1.RunFunctionResponse, error) {
```

Comment:
This is definitely an improvement, but I'm not usually a fan of taking a chunk of code and breaking it out into a private method, especially if it's only called in one place.

If I look at how this method uses its `*FunctionComposer`:

* It calls `c.pipeline.RunFunction` - it's also the only thing that calls that method.
* It uses `c.composite.ExtraResourcesFetcher` to fetch extra resources (if enabled)

This makes me wonder whether it could be refactored to become a new implementation of the existing `FunctionRunner` interface that takes an underlying `FunctionRunner` (e.g. the existing `PackagedFunctionRunner`) and wraps it to add extra resources support.

At a glance I think it'd be a pretty small change. You could use the extra resources feature flag to determine whether `Compose` should use a bare `PackagedFunctionRunner`, or one wrapped with extra resources support.

Reply:
Get your point, I'll be off for the rest of the week, I'd get this one merged as is and follow up with a refactor adding some tests too.

Reply:
I ended up deciding to "quickly" tackle it myself today: https://github.com/crossplane/crossplane/pull/5883. (Spoiler alert: it's never as quick as I think it'll be.)

Reply:
Thanks! Yeah, it never is. That's why I kind of liked the named loop fix. It was actually quick and concise.